sycl/doc/design/PropagateCompilerFlagsToRuntime.md

```diff
@@ -0,0 +1,66 @@
+
+Here is an example that demonstrates this pain point:
+
+```
+clang++ -c test_host.cpp -o test_host.o
+clang++ -c -fsycl test_device_1.cpp -o test_device_1.o
+clang++ -c -fsycl -O0 test_device_2.cpp -o test_device_2.o
+clang++ -fsycl -o test test_host.o test_device_1.o test_device_2.o
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test_device_2.c module, some of the debuggability is lost.
+
+Another scenario is shown below:
+
+```
+clang++ -c -O0 -fsycl -g test.cpp -o test.o
+clang++ -fsycl test.o -o test
+```
+
+In this scenario, the fat binary is 'test' and there are no compilation flags sent across to the backend compiler. Though the user wanted to have full debuggability with test.cpp module, some of the debuggability is lost. The user was not able to set a breakpoint inside device code.
+
+## Requirements
+
+In order to support module-level debuggability, the user will compile different module files with different levels of optimization. These optimization levels must be preserved and made use of during the backend compilation. Following are the requirements for this feature.
+- If the user specifies `-Ox` as a front-end compile option for a particular module, this option must be preserved during backend JIT compilation.
+- If the user specifies `-Ox` option using the environment variable, this option will override any front-end compile option and the new option will be preserved during JIT compilation.
+- If the user specifies `-O0` option, SYCL runtime needs to pass the appropriate backend option to JIT compilation stages.
+
+The following table specifies the appropriate backend options for level-zero and OpenCL backends.
+
+| Front-end option | L0 backend option | OpenCL backend option |
+| ---------------- | ----------------- | --------------------- |
+|      -O0         |  -ze-opt-disable  |   -cl-opt-disable     |
+|      -O1         |  -ze-opt-level=1  |   /* no option */     |
+|      -O2         |  -ze-opt-level=1  |   /* no option */     |
+|      -O3         |  -ze-opt-level=2  |   /* no option */     |
+
+
+## Proposed design
+
+This chapter discusses changes required in various stages of the compilation pipeline.
+
+
+### Changes to the DPC++ front-end
+
+For each SYCL kernel, we add a new function attribute that is named `sycl-optlevel`. Value of this attribute is set to the optimization level which was used to compile the overlying module.
```

Comment:
By "DPC++ front-end", I assume you mean some change to the CFE (i.e. before LLVM IR generation)?

I wonder if we want to continue making SYCL specific changes to the CFE.  Would it be better to add this IR attribute during an LLVM IR pass?  We already have an IR pass that runs early in the pipeline, which we use for the "optional kernel feature" design.  We could add the attribute during that pass instead.

Reply:
We want to capture the options that clang front end gets called with. So this is the appropriate place for this addition. I do agree I need to call it CFE. Thanks.

Reply:
I'm not that familiar with the clang structure, but I thought the IR passes and the CFE were all in the same process.  Aren't the command line options available to the IR passes?

Reply:
@AlexeySachkov, @steffenlarsen, can you please comment on which pass is used for "optional kernel features" and if I can use it here to add the attribute? 

Thanks

Reply:
Optimization level information is known at the time we are building pipeline, because we do that within `clang++` invocation, right after we emitted LLVM IR:

https://github.com/intel/llvm/blob/4480f8b279fa1fad5dbaa2f0ea39fb1cdcd02096/clang/lib/CodeGen/BackendUtil.cpp#L1010-L1018

The pass Greg is talking about is [`SYCLPropagateAspectsUsage`](https://github.com/intel/llvm/blob/sycl/llvm/lib/SYCLLowerIR/SYCLPropagateAspectsUsage.cpp).

I agree that adding the metadata in a pass would be a more modular solution which simplifies the design in context of amount of affected components.

Reply:
I have added a separate pass to add this attribute. Thanks